### PR TITLE
Assignment 5

### PR DESCRIPTION
**Pull request by c-stanton**
Worked with @Al3xramirez and added in extra credits for both ICollection<T> on Node class and the VennDiagram structure.

### CODE REVIEW

File: GenericsHomeWork.Tests/GenericsHomeWork.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <EnableNetAnalyzers>true</EnableNetAnalyzers>
```
Comment (ulises-aguilar) [inline GenericsHomeWork.Tests/GenericsHomeWork.Tests.csproj]: Place all shared project properties into a Directory.Build.Props file.

**Comment by JosephPotapenko** [comment]
Was about to say the same thing Ulises said. No Directory.Build.Props to fulfill the requirement:
 Place all shared project properties into a Directory.Build.Props file.

**Review by ulises-aguilar — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
Create an Exists method to test whether a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔

**Comment by JosephPotapenko** [comment]
You could add in checks in the add and contains methods to make sure the item passed in isn't null, something like this possibly:
Contains:
     return item != null && _items.Contains(item);
Add:
    ArgumentNullException.ThrowIfNull(item);
    _items.Add(item);

You could check that circle name doesn't come in null like Mark showed in class yesterday:
    Name = name ?? throw new ArgumentNullException(nameof(name));


Once you set up  Directory.Build.props from my previous comment, you can get rid of the duplicate code from GenericsHomeWork.csproj and GenericsHomeWork.Tests.csproj. 


Possibly do some null checks on the extra credit too if you want?

Most of my nitpicks are edge cases though, you seem to have gotten most everything else done well.

**Comment by JosephPotapenko** [comment]
## Instructions

- Create a *class library* project called "GenericsHomework.". ✔
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
  - Define the `Node` class
  - The values should be homogeneous.
  - Include a constructor that takes a value.  (No validation is necessary on the value). ✔
  - Add a `ToString()` override that writes out the value's `ToString()` result. ✔
  - Add a `Next` property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
    - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
    - The `Next` property setter should be private. ✔
  - Add an `Append` method that takes a value and appends a new `Node` instance after the current node (by invoking the `Next` property). ✔
  - Add a `Clear` method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
    - Whether it is sufficient to only set Next to itself ✔
    - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
    - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
  - Create an Exists method to test whether a value exists in the list. ✔
  - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

## Extra Credit

It looks like it was attempted. Didn't check it since it's not specifically required.

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file. ❌
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: GenericsHomeWork.Tests/GenericsHomeWork.Tests.csproj (hunk 1 above)
Comment (Al3xramirez) [reply GenericsHomeWork.Tests/GenericsHomeWork.Tests.csproj]: Thank you Ulises, will get that fixed

**Comment by kellanburns** [comment]
- Create a class library project called "GenericsHomework.". ❌ (not a 'class library')
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item. ✔
   - Define the Node class ✔
   - The values should be homogeneous. ✔
   - Include a constructor that takes a value. (No validation is necessary on the value). ✔
   - Add a ToString() override that writes out the value's ToString() result. ✔
   - Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
      - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
      - The `Next` property setter should be private. ✔
   - Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
   - Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
      - Whether it is sufficient to only set Next to itself ✔
      - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
      - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
   - Create an Exists method to test whether a value exists in the list. ✔
   - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit
1. Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ✔
- Each circle contains n items and each item can belong to one or more Circle instances.
- You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
- You are welcome to use existing BCL generic classes for the extra credit.
2. Implement System.Collections.Generic.ICollection<T> on the Node class ✔

Fundamentals
- Place all shared project properties into a Directory.Build.Props file.
nullable reference types is enabled ❌ (no .Props file)
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
- For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

Most of it looks good, previous reviews have already mentioned the issues with 'class library' and the missing Directory.Build.Props file. I also think the exception thrown by `Append()` should be an `ArgumentException`, because the provided `InvalidOperationException` is more relevant to an object's state than a generally disallowed operation (which more closely fits the parameters of an ArgumentException).

File: GenericsHomeWork/Node.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    public bool Contains(T item)
+    {
+        return Exists(item);
+    }
+
+    public void Append(T value)
+    {
+        if (Exists(value))
+        {
+            throw new InvalidOperationException($"Value '{value}' already exists in the list.");
```
Comment (kellanburns) [inline GenericsHomeWork/Node.cs]: I think the exception type should be changed from `InvalidOperationException` to `ArgumentException`, because the use-case for the InvalidOperation exception is when the current state of an object makes the operation invalid, not that the operation isn't allowed to happen. 

Also, a bit of a nitpick, formatting the exception with `nameof(value)` being passed after the message is more common and lets you describe the 'no duplicates allowed', versus just alerting about the duplicate. 

```suggestion
            throw new ArgumentException("No duplicate values allowed in the list.", nameof(value));
```

**Review by kellanburns — CHANGES_REQUESTED** [review]

**Comment by c-stanton** [comment]
> * s

It defaults to a class library when there is an absence of a Main. The build gives it a .dll extension, and when I view the properties of the project it shows that it is a class library: 
default
generics1
generics2

File: GenericsHomeWork/Node.cs (hunk 2 above)
Comment (c-stanton) [reply GenericsHomeWork/Node.cs]: I changed it, thank you for your suggestion!

**Review by katmilton — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
Create an Exists method to test whether a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔


Extra Credit

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ✔
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection<T> on the Node class ✔


Fundamentals

Place all shared project properties into a Directory.Build.Props file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔


Great work!

File: GenericsHomeWork/Node.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        }
+        this.Next = this;
+    }
+
+    public int Count
+    {
+        get
+        {
+            int count = 0;
+            NodeCollection<T>? currNode = this.Next;
```
Comment (Joshua-Lester3) [inline GenericsHomeWork/Node.cs:74]: I'm not sure why this is a nullable variable, since Next is non-nullable. Also, I'd avoid using `this.` as it's not needed here.

**Review by Joshua-Lester3 — COMMENTED** [review]
## Instructions

- Create a _class library_ project called "GenericsHomework.". ❌ Not specified in csproj
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
  - Define the `Node` class
  - The values should be homogeneous.
  - Include a constructor that takes a value. (No validation is necessary on the value). ✔
  - Add a `ToString()` override that writes out the value's `ToString()` result. ✔
  - Add a `Next` property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
    - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
    - The `Next` property setter should be private. ✔
  - Add an `Append` method that takes a value and appends a new `Node` instance after the current node (by invoking the `Next` property). ✔
  - Add a `Clear` method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
    - Whether it is sufficient to only set Next to itself ✔
    - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
    - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
  - Create an Exists method to test whether a value exists in the list. ✔
  - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

1. Implement a `VennDiagram` structure that contains `n` `Circle`s that only contains homogenous **reference types** of any type. ✔

- Each circle contains n items and each item can belong to one or more `Circle` instances.
- You are not required to use a `Node` from earlier in the homework for your Venn diagram implementation.
- You are welcome to use existing BCL generic classes for the extra credit.

1. Implement `System.Collections.Generic.ICollection<T>` on the `Node` class ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version) ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

**Comment by Joshua-Lester3** [comment]
> It defaults to a class library when there is an absence of a Main. The build gives it a .dll extension, and when I view the properties of the project it shows that it is a class library:

I think it defaults to this if there's no Main, but specifying it explicitly avoids ambiguity.
